File: Kotlin/src/PerformanceAnalyze.kt
import Benchmarks.Fasta
import Benchmarks.Knucleotide
import kotlin.system.measureTimeMillis
import java.io.File
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

private const val fastaOutputPath = "fasta.txt"

fun main() {
    PerformanceAnalyze.Run("FannkuchRedux", 10) { Benchmarks.FannkuchRedux.runBenchmark(12) }
    PerformanceAnalyze.Run("Nbody", 10) { Benchmarks.Nbody.runBenchmark(50000000) }
    PerformanceAnalyze.Run("Fasta", 10) { Fasta.runBenchmark(fastaOutputPath, 25000000) }
    PerformanceAnalyze.Run("Knucleotide", 10) { Knucleotide.runBenchmark(fastaOutputPath) }

    val myFile = File(fastaOutputPath)
    if (myFile.exists())
        myFile.delete()
}

class PerformanceAnalyze {
    companion object {
        fun Run(name: String, iterations: Int, action: () -> Unit) {

            // Perform garbage collection.
            System.gc()
            System.runFinalization();

            // Force JIT compilation of the method.
            action()

            print("Time '${LocalDateTime.now().format(DateTimeFormatter.ISO_TIME)}'. Running benchmark '${name}' for $iterations iterations... ")

            var time = 0L;
            for (i in 1..iterations) {
                print("$i ")
                time += measureTimeMillis {
                    action()
                }
            }
            println("| Time '${LocalDateTime.now().format(DateTimeFormatter.ISO_TIME)}'. Elapsed time ${time / iterations} ms.")
        }
    }
}

File: Csharp/PerformanceAnalyze.cs
using System;
using System.Diagnostics;
using System.IO;
using Benchmarks;

namespace PerformanceAnalyze
{
    public static class PerformanceAnalyze
    {
        private const string fastaOutputPath = @"fasta.txt";

        public static void Main(string[] args)
        {
            Run("FannkuchRedux", 10, () => FannkuchRedux.RunBenchmark(12));
            Run("Nbody", 10, () => NBody.RunBenchmark(50000000));
            // Run Fasta before KNucleotide, because it uses his output
            Run("Fasta", 10, () => Fasta.RunBenchmark(fastaOutputPath, 25000000));
            Run("KNucleotide", 10, () => KNucleotide.RunBenchmark(fastaOutputPath));

            if (File.Exists(fastaOutputPath))
            {
                File.Delete(fastaOutputPath);
            }
        }

        public static void Run(string name, int iterations, Action action)
        {
            Console.Write($"Running benchmark '{name}' for {iterations} iterations... ");

            // Perform garbage collection.
            GC.Collect();
            GC.WaitForPendingFinalizers();

            // Force JIT compilation of the method.
            action.Invoke();

            // Run the benchmark.
            Stopwatch watch = Stopwatch.StartNew();
            for (int i = 0; i < iterations; i++)
            {
                Console.Write($"{i + 1} ");
                action.Invoke();
            }

            watch.Stop();

            // Output results.
            Console.WriteLine($"| Elapsed time {watch.ElapsedMilliseconds / iterations} ms.");
        }
    }
}